feat(web-analytics): Add Web Vitals to the toolbar

rafaeelaudibert opened the pull request:
We can now look at web vitals from our toolbar. It'll display information from the current page run but also from the last 7 days in that specific page.

![image](https://github.com/user-attachments/assets/23271ae5-7d60-422a-b174-053cad3fb86e)

This is hidden behind a separate FF so this should be pretty safe to merge

rafaeelaudibert committed:
feat: Allow spinner speed to be customizable

We can customize the spinner speed now, just pass a "${number}s" string, it's 1s by default

rafaeelaudibert committed:
feat(web-analytics): Add new `web-vitals-toolbar` FF

We'll be adding a new Web Vitals section to the toolbar, but let's keep it hidden behind a FF

rafaeelaudibert committed:
refactor: Move functions to `definitions.ts`

We had some leftovers functions inside `webAnalyticsLogic`, let's keep them inside `nodes/WebVitals/definitions.ts` instead, it makes more sense and simplifies the whole import/export dance

rafaeelaudibert committed:
feat: Add new `/web_vitals` endpoint

This endpoint can be used to collect information from the most recent 7 days of data for your Web Vitals, this is used in the toolbar

rafaeelaudibert committed:
chore: Add "no typegen" build for VSCode

I'm using this locally, it helps with my workflow, let's stop having this in my git stash all the time

rafaeelaudibert committed:
feat: Add Web Vitals to the toolbar

We can now look at web vitals from our toolbar. It'll display information from the current page run but also from the last 7 days in that specific page.

rafaeelaudibert committed:
feat: Use cleaned path operator

This will return more reasonable values to the UI

rafaeelaudibert committed:
refactor: Use cached value if existent on `web_vitals.py`

The previous value was better for local testing, but we should use the
cached value in production

rafaeelaudibert committed:
fix: Use new version of function

After rebasing on master we had to fix this

rafaeelaudibert commented on frontend/src/lib/lemon-ui/Spinner/Spinner.tsx:
```diff
@@ -6,12 +6,15 @@ import { twJoin, twMerge } from 'tailwind-merge'
     return (
         <svg
+            // eslint-disable-next-line react/forbid-dom-props
+            style={{ '--speed': speed } as React.CSSProperties}
```
I need the variable, so this needs to be like this

rafaeelaudibert committed:
fix: typo

rafaeelaudibert commented on frontend/src/lib/constants.tsx:
```diff
@@ -235,6 +235,7 @@ export const FEATURE_FLAGS = {
+    WEB_VITALS_TOOLBAR: 'web-vitals-toolbar', // owner: @rafaeelaudibert #team-web-analytics
     WEB_REVENUE_TRACKING: 'web-revenue-tracking', // owner: @robbie-c #team-web-analytics
     LLM_OBSERVABILITY: 'llm-observability', // owner: #team-ai-product-manager
     ONBOARDING_SESSION_REPLAY_SEPERATE_STEP: 'onboarding-session-replay-separate-step', // owner: @joshsny #team-growth
```
Not my code, but I'll fix it :)

rafaeelaudibert commented on frontend/src/scenes/settings/user/personalAPIKeysLogic.tsx:
```diff
@@ -99,6 +99,7 @@ export const APIScopes: APIScope[] = [
         },
     },
     { key: 'webhook', info: 'Webhook configuration is currently only enabled for the Zapier integration.' },
+    { key: 'web_vitals' },
```
@zlwaterfield that's where I need you. Do I need to do anything here?

zlwaterfield replied:
I haven't worked much with the api scopes, what is your goal? Does the toolbar load data through personal API keys?

rafaeelaudibert replied:
I thought this was RBAC related, so that's why I tagged you, my apologies hahah

It uses the same `temporary_token` other toolbar stuff is using, so I don't think I'm using personal API Keys

rafaeelaudibert commented on frontend/src/toolbar/ToolbarApp.tsx:
```diff
@@ -18,6 +19,13 @@ export function ToolbarApp(props: ToolbarProps = {}): JSX.Element {
+    // Preemptively mount the web vitals toolbar logic on mount
+    // so that we collect the web vitals metrics since the beginning
+    // TODO: Should probably mount the event debugger logic here too
+    useEffect(() => {
+        webVitalsToolbarLogic.mount()
+    }, [])
```
Intentional, keep collecting metrics

rafaeelaudibert committed:
style: Simplify scss rule

rafaeelaudibert committed:
fix: Remove FF inconsistency

rafaeelaudibert commented on frontend/src/toolbar/bar/Toolbar.tsx:
```diff
@@ -94,9 +96,16 @@ export function ToolbarInfoMenu(): JSX.Element | null {
+
+    const showWebVitalsFlag = useToolbarFeatureFlag('web-vitals')
+    const showWebVitalsToolbarFlag = useToolbarFeatureFlag('web-vitals-toolbar')
+    const showWebVitals = showWebVitalsFlag && showWebVitalsToolbarFlag
```
Fixed

rafaeelaudibert commented on frontend/src/toolbar/bar/Toolbar.tsx:
```diff
@@ -154,8 +165,13 @@ export function Toolbar(): JSX.Element | null {
+    const showExperiments = inStorybook() || inStorybookTestRunner() || showExperimentsFlag
+
+    const showWebVitalsFlag = useToolbarFeatureFlag('web-vitals')
+    const showWebVitalsToolbarFlag = useToolbarFeatureFlag('web-vitals-toolbar')
+    const showWebVitals = inStorybook() || inStorybookTestRunner() || (showWebVitalsFlag && showWebVitalsToolbarFlag)
```
Hard because it's the toolbar, in a real world would make sense

rafaeelaudibert commented on frontend/src/toolbar/web-vitals/WebVitalsToolbarMenu.tsx:
```diff
@@ -0,0 +1,133 @@
+            >
+                {metric}:
+            </DottedTooltip>
+            <span className={clsx('text-sm', `text-${color}`)}>
```
Yes, and that's why they're strictly validated, we control that function output

rafaeelaudibert commented on posthog/api/__init__.py:
```diff
@@ -589,3 +590,10 @@ def register_grandfathered_environment_nested_viewset(
     r"data_color_themes", data_color_theme.DataColorThemeViewSet, "environment_data_color_themes", ["team_id"]
 )
+
+projects_router.register(
+    r"web_vitals",
+    web_vitals.WebVitalsViewSet,
+    "project_web_vitals",
+    ["project_id"],
+)
```
Do I need to support both here @Twixes ?

Twixes replied:
Tackling in greptile's other comment on this line

rafaeelaudibert commented on posthog/api/web_vitals.py:
```diff
@@ -0,0 +1,79 @@
+        if not request.user.is_authenticated:  # for mypy
+            raise exceptions.NotAuthenticated()
+
+        pathname = request.query_params.get("pathname")
```
I assume Django will handle this for me automatically, won't it?

Twixes replied:
Nope, because Django REST Framework can only handle validation for you if you're using a serializer on the viewset. By defining `list()` yourself, you're promising to take care of the validation yourself

rafaeelaudibert commented on posthog/api/web_vitals.py:
```diff
@@ -0,0 +1,79 @@
+from posthog.hogql_queries.query_runner import get_query_runner, ExecutionMode
+
+
+class WebVitalsViewSet(TeamAndOrgViewSetMixin, viewsets.ModelViewSet):
```
Question to those with more REST Django knowledge than I have (nil), do I actually need to inherit anything here?

Twixes replied:
Because you're not actually representing a Django model with this viewset, you should not be inheriting from `ModelViewSet` – just `GenericViewSet`

rafaeelaudibert committed:
refactor: Use more restrictive viewset

We never write, so let's remove some endpoints we don't need

rafaeelaudibert committed:
refactor: Throw manual 500 on missing data

rafaeelaudibert committed:
fix: Import without tsx

rafaeelaudibert committed:
tests: Add missing FF for Toolbar snapshots

github-actions[bot] committed:
Update query snapshots

rafaeelaudibert committed:
fix: Avoid rendering Spinner on storybook

This caused our storybook tests to never finish running, see https://github.com/PostHog/posthog/actions/runs/13085431533/job/36515998085?pr=28173

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

rafaeelaudibert committed:
tests: Add some color to tests

Make some of the metrics yellow/red

rafaeelaudibert commented on frontend/src/lib/constants.tsx:
```diff
@@ -235,9 +235,10 @@ export const FEATURE_FLAGS = {
     WEB_REVENUE_TRACKING: 'web-revenue-tracking', // owner: @robbie-c #team-web-analytics
     LLM_OBSERVABILITY: 'llm-observability', // owner: #team-ai-product-manager
-    ONBOARDING_SESSION_REPLAY_SEPERATE_STEP: 'onboarding-session-replay-separate-step', // owner: @joshsny #team-growth
+    ONBOARDING_SESSION_REPLAY_SEPARATE_STEP: 'onboarding-session-replay-separate-step', // owner: @joshsny #team-growth
```
@joshsny Updated this because greptile was complaining about it, hope you don't mind

Twixes commented on frontend/src/lib/lemon-ui/Spinner/Spinner.scss:
```diff
@@ -1,4 +1,5 @@
 .Spinner {
+    --speed: 1s;
```
Let's use `--spinner-speed` for clarity, matching `--spiner-color`

rafaeelaudibert replied:
100%

rafaeelaudibert replied:
Done, see 0bcdf527e217561c2cddac7ec759b8268ce61b9f

Twixes commented on posthog/api/__init__.py:
```diff
@@ -589,3 +590,10 @@ def register_grandfathered_environment_nested_viewset(
     r"data_color_themes", data_color_theme.DataColorThemeViewSet, "environment_data_color_themes", ["team_id"]
 )
+
+projects_router.register(
+    r"web_vitals",
+    web_vitals.WebVitalsViewSet,
+    "project_web_vitals",
+    ["project_id"],
+)
```
@rafaeelaudibert This is a great point from Greptile, i's absolutely right that there shouldn't be a mismatch.
Answering your other question at the same time: this viewset must be registered under `environments_router`, since it returns query results i.e. data, and data always is environment-level.
As a new endpoint, it should _only_ be registered under `environments_router`, and should only be using `team_id`.

rafaeelaudibert replied:
Nice, thanks, see b8b6b680fe642a81425706f901cebaf7582e2f49

rafaeelaudibert committed:
refactor: Rename var to `--spinner-speed`

Align with the existing `--spinner-color` var

rafaeelaudibert committed:
fix: Properly register new endpoint under `environments_router`

Per https://github.com/PostHog/posthog/pull/28173#discussion_r1939319387

github-actions[bot] committed:
Update query snapshots

Twixes commented on frontend/src/toolbar/Toolbar.stories.tsx:
```diff
@@ -81,6 +84,7 @@ const BasicTemplate: StoryFn<ToolbarStoryProps> = (props) => {
             '/api/projects/@current/feature_flags/my_flags': listMyFlagsAPIResponse,
             '/api/projects/@current/actions/': listActionsAPIResponse,
             '/api/projects/@current/web_experiments/': listExperimentsAPIResponse,
+            '/api/projects/@current/web_vitals/': listWebVitalsAPIResponse,
```
So this will now be `/environments/`

Twixes commented on frontend/src/toolbar/web-vitals/webVitalsToolbarLogic.ts:
```diff
@@ -0,0 +1,126 @@
+                getWebVitals: async (_, breakpoint) => {
+                    const params = { pathname: window.location.pathname }
+
+                    const response = await toolbarFetch(`/api/projects/@current/web_vitals${encodeParams(params, '?')}`)
```
`/environments/`

Twixes commented on posthog/api/web_vitals.py:
```diff
@@ -0,0 +1,79 @@
+
+
+class WebVitalsViewSet(TeamAndOrgViewSetMixin, viewsets.ReadOnlyModelViewSet):
+    scope_object = "web_vitals"
```
Let's instead use the existing `scope_object = "query"` value, since this endpoint just boils down to a query which can be used with `TemporaryTokenAuthentication`.

But wait – it doesn't looks like `TemporaryTokenAuthentication` is actually allowed on this viewset, so this would in fact not work for PostHog users (it only works with PostHog in local dev's self-capture mode)

rafaeelaudibert replied:
Ok, good catch, I wasn't aware of that

I've moved to using `query` here and added a new authentication class to allow the temporary token to work

rafaeelaudibert committed:
fix: Handle required `pathname` parameter

rafaeelaudibert committed:
fix: Inherit from `GenericViewSet`

We're not inheriting from a Django model or anything, so let's use the
generic viewset instead

rafaeelaudibert committed:
fix: Use correct `/environments` endpoint

rafaeelaudibert committed:
refactor: Use `query` scope for WebVitals toolbar API endpoint

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

Twixes reviewed (commented):
I don't see blockers anymore, but gonna let @robbie-c give the final ✅, since you two are much closer working on web analytics 😄

robbie-c commented on frontend/src/queries/nodes/WebVitals/definitions.ts:
```diff
@@ -72,10 +72,9 @@ export const getMetric = (
-    threshold: WebVitalsThreshold
-): WebVitalsMetricBand | 'none' => {
+export const getMetricBand = (value: number | undefined, metric: WebVitalsMetric): WebVitalsMetricBand | 'none' => {
+    const threshold = WEB_VITALS_THRESHOLDS[metric]
```
👍 to moving this lookup here (and below in `getThresholdColor`), I think it makes everything that calls it feel cleaner

robbie-c commented on frontend/src/toolbar/bar/Toolbar.scss:
```diff
@@ -139,10 +139,15 @@
         --toolbar-width-expanded: 334px;
     }
 
+    &--with-experiments-and-web-vitals {
```
hmm, this approach is probably fine for now but seems unscalable 😅

robbie-c commented on posthog/api/web_vitals.py:
```diff
@@ -0,0 +1,85 @@
+
+# This is a simple wrapper around a basic query, so that's why `scope_object = "query"`
+# This `Viewset` does need to exist, however, because we need to support the `TemporaryTokenAuthentication` method
+class WebVitalsViewSet(TeamAndOrgViewSetMixin, viewsets.GenericViewSet):
```
I don't feel super able to review this part, I've not touched this part of django, but if @Twixes is happy then so am I

robbie-c commented on posthog/models/scopes.py:
```diff
@@ -1,9 +1,3 @@
-## API Scopes
```
Why get rid of the comment?

rafaeelaudibert replied:
It's duplicated below

pauldambra commented:
this is cool... now i want to be able to get this data for prod while i have the toolbar on my local dev env!
